File: artifacts/api-server/src/lib/bagWorker.ts
/**
 * bagWorker.ts — Persistent Python subprocess for BAG file parsing.
 *
 * Keeps one `bag_worker.py` process alive across all parseBag() calls so
 * that Python + h5py + pyproj are loaded only once (~500–700 ms cold start)
 * instead of on every invocation.
 *
 * Protocol (stdin/stdout, line-oriented):
 *   Send:    <absolute path to .bag tmp file>\n
 *   Receive: \n<csv lines>__OK__\n   on success
 *            \n__ERR__\t<msg>\n      on failure (embedded \n escaped as \\n)
 *
 * Concurrent calls are serialised through an internal queue; the worker
 * processes one BAG file at a time.
 *
 * Cross-file singleton lifetime
 * ─────────────────────────────
 * The BagWorkerProcess instance is stored under a well-known global symbol
 * (WORKER_GLOBAL_KEY).  This allows the same instance — and thus the same
 * Python child process — to be reused even when vitest re-imports this module
 * for each test file (module isolation mode).  Combined with vitest's
 * `poolOptions.forks.singleFork: true`, all test files run inside the same
 * forked OS process and the global object is truly shared, giving every BAG
 * test file access to the one warm Python worker.
 */

import { spawn, type ChildProcess } from "child_process";
import { existsSync } from "fs";
import { dirname, join } from "path";
import { fileURLToPath } from "url";
import { logger } from "./logger.js";

// ---------------------------------------------------------------------------
// Script / env resolution
// ---------------------------------------------------------------------------

function findBagWorkerScript(): string {
  const thisDir = dirname(fileURLToPath(import.meta.url));
  const candidates = [
    join(thisDir, "bag_worker.py"),       // dev: src/lib/  |  prod main: dist/
    join(thisDir, "..", "bag_worker.py"), // prod worker: dist/lib/ → dist/
  ];
  for (const c of candidates) {
    if (existsSync(c)) return c;
  }
  throw new Error(
    `bag_worker.py not found. Searched: ${candidates.join(", ")}. ` +
      "Rebuild the server with `pnpm build` to copy the script to dist/.",
  );
}

function findPythonUserBase(scriptPath: string): string | undefined {
  let dir = dirname(scriptPath);
  for (let i = 0; i < 6; i++) {
    const candidate = join(dir, ".pythonlibs");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  return undefined;
}

// ---------------------------------------------------------------------------
// Internal queue item
// ---------------------------------------------------------------------------

interface QueuedRequest {
  path: string;
  resolve: (csv: string) => void;
  reject: (err: Error) => void;
}

// ---------------------------------------------------------------------------
// BagWorkerProcess
// ---------------------------------------------------------------------------

/**
 * Manages a single persistent `bag_worker.py` subprocess.
 * Spawned lazily on first use; restarts automatically after an unexpected exit.
 */
class BagWorkerProcess {
  private proc: ChildProcess | null = null;
  private stdoutBuf = "";
  private active: { resolve: (csv: string) => void; reject: (err: Error) => void } | null = null;
  private queue: QueuedRequest[] = [];

  /**
   * Parse a BAG file at `bagPath` (absolute path to a temp file).
   * Returns the raw CSV string (lon,lat,depth rows).
   */
  parseFile(bagPath: string): Promise<string> {
    return new Promise<string>((resolve, reject) => {
      if (this.active === null) {
        this.active = { resolve, reject };
        this._ensureProc().stdin!.write(bagPath + "\n");
      } else {
        this.queue.push({ path: bagPath, resolve, reject });
      }
    });
  }

  /**
   * Gracefully shut down the worker process.
   * Closing stdin signals EOF to the Python worker, which exits cleanly.
   * In tests, prefer letting the OS process exit handle cleanup (via unref)
   * so the shared worker stays alive across multiple test files.
   */
  shutdown(): void {
    const p = this.proc;
    this.proc = null;
    if (p) {
      try {
        p.stdin?.end();
      } catch {
        // ignore
      }
    }
  }

  // -------------------------------------------------------------------------
  // Private helpers
  // -------------------------------------------------------------------------

  private _ensureProc(): ChildProcess {
    if (this.proc) return this.proc;

    const scriptPath = findBagWorkerScript();
    const pythonUserBase = findPythonUserBase(scriptPath);
    const env: NodeJS.ProcessEnv = {
      ...process.env,
      ...(pythonUserBase ? { PYTHONUSERBASE: pythonUserBase } : {}),
    };

    const proc = spawn("python3", [scriptPath], {
      env,
      stdio: ["pipe", "pipe", "pipe"],
    });

    // Don't keep the Node process alive just because the worker is running.
    proc.unref();

    // Spawn failure (e.g. python3 not found) — reject the in-flight request
    // instead of emitting an unhandled 'error' event that would crash Node.
    proc.on("error", (spawnErr) => {
      this.proc = null;
      const cur = this.active;
      const pending = [...this.queue];
      this.active = null;
      this.queue = [];
      this.stdoutBuf = "";
      const err = new Error(`BAG parse error: failed to spawn python3 — ${spawnErr.message}`);
      if (cur) cur.reject(err);
      for (const q of pending) q.reject(err);
    });

    proc.stdout!.on("data", (chunk: Buffer) => {
      this.stdoutBuf += chunk.toString("utf8");
      this._drain();
    });

    proc.stderr!.on("data", (chunk: Buffer) => {
      const text = chunk.toString("utf8").trim();
      if (text && text !== "bag_worker: ready") {
        logger.warn({ source: "bag_worker.py" }, text);
      }
    });

    proc.on("exit", (code, signal) => {
      this.proc = null;
      // Reject any in-flight or queued request so callers don't hang.
      const cur = this.active;
      const pending = [...this.queue];
      this.active = null;
      this.queue = [];
      this.stdoutBuf = "";
      if (cur) {
        cur.reject(
          new Error(
            `bag_worker.py exited unexpectedly (code=${code}, signal=${signal})`,
          ),
        );
      }
      for (const q of pending) {
        q.reject(
          new Error(
            `bag_worker.py exited unexpectedly (code=${code}, signal=${signal})`,
          ),
        );
      }
    });

    this.proc = proc;
    return proc;
  }

  /** Process any complete response frames in stdoutBuf. */
  private _drain(): void {
    const OK_MARKER = "\n__OK__\n";
    const ERR_MARKER = "\n__ERR__\t";

    while (this.active) {
      const okIdx = this.stdoutBuf.indexOf(OK_MARKER);
      const errIdx = this.stdoutBuf.indexOf(ERR_MARKER);

      // Determine which (if any) marker appears first.
      let which: "ok" | "err" | null = null;
      if (okIdx !== -1 && (errIdx === -1 || okIdx <= errIdx)) which = "ok";
      else if (errIdx !== -1) which = "err";

      if (which === "ok") {
        const csv = this.stdoutBuf.slice(0, okIdx);
        this.stdoutBuf = this.stdoutBuf.slice(okIdx + OK_MARKER.length);
        const cur = this.active;
        this.active = null;
        this._next();
        cur.resolve(csv);
      } else if (which === "err") {
        const afterTag = this.stdoutBuf.slice(errIdx + ERR_MARKER.length);
        const nlIdx = afterTag.indexOf("\n");
        if (nlIdx === -1) break; // incomplete line — wait for more data
        const rawMsg = afterTag.slice(0, nlIdx);
        this.stdoutBuf = afterTag.slice(nlIdx + 1);
        const cur = this.active;
        this.active = null;
        this._next();
        // Restore escaped newlines so error messages remain readable.
        const msg = rawMsg.replace(/\\n/g, "\n");
        cur.reject(new Error(`BAG parse error: ${msg}`));
      } else {
        break; // no complete frame yet
      }
    }
  }

  /** Dispatch the next queued request, if any. */
  private _next(): void {
    if (this.queue.length === 0) return;
    const next = this.queue.shift()!;
    this.active = { resolve: next.resolve, reject: next.reject };
    this._ensureProc().stdin!.write(next.path + "\n");
  }
}

// ---------------------------------------------------------------------------
// Cross-file singleton via global symbol
// ---------------------------------------------------------------------------
//
// Storing the instance under a well-known Symbol.for key ensures that even
// when vitest re-imports this module for each test file (module isolation),
// the same BagWorkerProcess — and the same Python child process — is reused
// for the entire vitest run (all files share one OS process via singleFork).

const WORKER_GLOBAL_KEY = Symbol.for("bathyscan.bagWorkerProcess");

function getOrCreateWorker(): BagWorkerProcess {
  const g = globalThis as Record<symbol, BagWorkerProcess | undefined>;
  if (!g[WORKER_GLOBAL_KEY]) {
    g[WORKER_GLOBAL_KEY] = new BagWorkerProcess();
  }
  return g[WORKER_GLOBAL_KEY]!;
}

/** Singleton persistent worker — shared across all parseBag() calls. */
export const bagWorker = getOrCreateWorker();

File: artifacts/api-server/src/lib/bag_worker.py
#!/usr/bin/env python3
"""
bag_worker.py — Persistent stdin/stdout worker for BAG file parsing.

Stays alive between calls so Python + h5py + pyproj are loaded only once,
eliminating per-request cold-start overhead (~500–700 ms each).

Protocol (one request/response cycle per file):
  Stdin  (one line):  <absolute path to .bag file>\\n
                      __WARMUP__\\n  (no-op ping — imports h5py, sends __WARMUP_OK__)
  Stdout (response):
    \\n                          separator (always present, aids marker search)
    <lon,lat,depth lines>       CSV (may be empty on error path)
    __OK__\\n                    on success  OR
    __ERR__\\t<msg>\\n            on failure  (\\n in msg replaced by \\\\n)
    __WARMUP_OK__\\n             on __WARMUP__ request

The worker reads until EOF then exits cleanly.
"""

import sys
import os
import traceback

# Allow importing bag_parser helpers from the same directory.
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

try:
    from bag_parser import parse_standard_bag, parse_vr_bag, MAX_POINTS  # type: ignore
except ImportError as _import_err:
    # Surface the error at request time rather than at startup so the
    # caller's stderr-based error path still works.
    def parse_standard_bag(*_a, **_kw):  # type: ignore
        raise ImportError(_import_err)
    def parse_vr_bag(*_a, **_kw):  # type: ignore
        raise ImportError(_import_err)
    MAX_POINTS = 2_000_000

# Import h5py at module scope so the cost is paid once on startup, not per
# request.  The warmup() call on the TypeScript side sends __WARMUP__ which
# triggers this import path and receives __WARMUP_OK__ as an ACK.
try:
    import h5py  # type: ignore  # noqa: F401 — import triggers JIT compilation
    _H5PY_AVAILABLE = True
    _H5PY_ERR: str | None = None
except ImportError as _h5py_err:
    _H5PY_AVAILABLE = False
    _H5PY_ERR = (
        f"Missing Python dependency: {_h5py_err}. "
        "Run: PYTHONUSERBASE=.pythonlibs pip install h5py numpy pyproj --user"
    )


def _send_err(msg: str) -> None:
    one_line = msg.replace("\n", "\\n").replace("\r", "")
    sys.stdout.write(f"\n__ERR__\t{one_line}\n")
    sys.stdout.flush()


def handle_warmup() -> None:
    """Respond to a __WARMUP__ ping — h5py is already imported at module level."""
    sys.stdout.write("__WARMUP_OK__\n")
    sys.stdout.flush()


def handle_one(path: str) -> None:
    if not _H5PY_AVAILABLE:
        _send_err(_H5PY_ERR or "h5py not available")
        return

    try:
        with h5py.File(path, "r") as f:
            if "BAG_root" not in f:
                raise ValueError("Not a valid BAG file: missing BAG_root group")

            bag = f["BAG_root"]

            meta_xml = ""
            if "metadata" in bag:
                raw = bag["metadata"][()]
                if isinstance(raw, bytes):
                    meta_xml = raw.decode("utf-8", errors="replace")
                elif hasattr(raw, "dtype") and raw.dtype.kind in ("S", "U", "O"):
                    if raw.dtype.kind == "S":
                        meta_xml = b"".join(raw.flat).decode("utf-8", errors="replace")
                    else:
                        meta_xml = "".join(str(c) for c in raw.flat)
                else:
                    meta_xml = str(raw)

            is_vr = "varres_refinements" in bag and "varres_metadata" in bag
            if is_vr:
                points = parse_vr_bag(bag, meta_xml, MAX_POINTS)
            else:
                points = parse_standard_bag(bag, meta_xml, MAX_POINTS)

        if not points:
            _send_err("BAG file produced no valid depth points.")
            return

        sys.stdout.write("\n")
        for lon, lat, depth in points:
            sys.stdout.write(f"{lon:.8f},{lat:.8f},{depth:.4f}\n")
        sys.stdout.write("__OK__\n")
        sys.stdout.flush()

    except Exception as exc:
        tb = traceback.format_exc()
        _send_err(f"{exc}\n{tb}")


def main() -> None:
    sys.stderr.write("bag_worker: ready\n")
    sys.stderr.flush()

    for line in sys.stdin:
        path = line.rstrip("\n\r")
        if not path:
            continue
        if path == "__WARMUP__":
            handle_warmup()
        else:
            handle_one(path)


if __name__ == "__main__":
    main()
